[QUZ-78][FEATURE] 중복 로그인 방지 및 유저 검증 추가

HMWG opened the pull request:
<!-- PR의 제목은 "[Jira 티켓 코드] [종류(ex.FEATURE)] 이슈이름" 으로 작성 -->

## ✨ 구현 기능 명세
- 중복 로그인 방지 기능 추가
해당 회원이 이미 로그인 중 이라면 로그인을 못하게
토큰 만료 or 로그아웃을 해야 로그인(토큰 발급) 가능

- 유저 검증
WebClient 사용해 유저 서비스에서 회원 검증
(게이트웨이는 webflex 환경이라 feign보다 WebClient가 더 안정적이라고 하네요)
회원은 redis에서 캐시로 조회 ( 캐시에 없을 시 db에 한번 더 조회 )

<!-- 해당 이슈에서 구현한 기능을 간단하게 작성 -->

## ✅ PR Point
user-service로 postman 테스트는 했는데 나중에 다른 서비스에서 잘 동작하는지 확인 부탁 드립니다
(gateway에서 토큰 검증 받으려면 user-service 서버도 돌아가야함)

로그인 정책도 제가 임의로 생각했는데 더 나은 방식이 있다면 알려주시면 감사하겠습니다!

<!-- 코드리뷰를 받고 싶은 부분, 새로운 지식을 얻게 된 부분 등등 공유하고 싶은 점을 작성 -->

## 😭 어려웠던 점
> swagger 설정은 처음이라 진행하다가 일단 뒤로 넘겼습니다 ㅜㅜ
> 최대한 빨리 만들게요

일단 swagger 설정을 긁어오긴 했는데 잘 될 지는 모르겠습니다 develop이랑 병합해서 테스트해보고 안되면 다시 작성해볼게요


<!-- 해결하기 어려웠거나 시간이 오래 걸린 부분 작성 -->

HMWG committed:
feat: 로그인 시 사용자 상태 저장 QUZ-78

# Conflicts:
#	user-service/user-domain/src/main/kotlin/com/grepp/quizy/user/domain/user/RedisTokenRepository.kt
#	user-service/user-domain/src/main/kotlin/com/grepp/quizy/user/domain/user/UserReader.kt
#	user-service/user-domain/src/main/kotlin/com/grepp/quizy/user/domain/user/exception/UserErrorCode.kt

HMWG committed:
feat: 중복 로그인 불허용 기능 추가 QUZ-78

HMWG committed:
feat: 게이트웨이 유저 검증 기능 추가 QUZ-78

HMWG committed:
feat: /api-docs 토큰 검증 제외 QUZ-78

HMWG committed:
refactor: 유저 검증 base url 수정 QUZ-78

HMWG committed:
refactor: api-docs 허용 변경 QUZ-78

HMWG committed:
merge: 병합 충돌 해결 QUZ-78

HMWG committed:
feat: gateway swagger 설정 추가 QUZ-78

JiwonKKang reviewed (approved):
고생하셨습니다!
저는 `유저 캐시` 와 `로그인 세션` 이 user-service가 아니라 gateway-service에 있으면 어떨까하는 생각입니다!

gateway-service에서 redis를 통해 위 2개를 관리하면 네트워크 비용이 훨씬 줄어들거같아요. 처음에는 user-service에 요청을 통해 유저를 조회해서 실제 존재하는지 검증한뒤에 gateway에서 캐싱을합니다. 이렇게하면 유저 서비스를 거치지않고 바로 조회할수있어서 비용이 절약될거같아요!

또한 로그인 세션도 인증/인가와 더욱 관련이 깊은거 같아 gateway에 좋을거같다고 생각하는데 어떻게 생각하시나요?

HMWG committed:
feat: gateway에서 캐싱된 유저를 먼저 확인하도록 수정 QUZ-78

NaMinhyeok reviewed (approved):
고생하셨습니다 !

s0o0bn reviewed (approved):
오 좋습니당 고생하셨습니다!! 👍

HMWG committed:
feat: gateway filter에서 허용된 엔드포인트도 헤더를 보내도록 수정 QUZ-78

- 토큰이 존재할 경우 토큰을 검증
- 토큰이 검증되지 않으면 예외
- 토큰이 없을 시 허용된 엔드포인트면 패스
- 토큰이 없고 불허된 엔드포인트면 예외
